www/blog/2022-11-23-library-ts-perf.md

```diff
@@ -0,0 +1,197 @@
+author_title: tRPC Core Team Member
+author_url: https://twitter.com/s4chinraja
+author_image_url: https://avatars1.githubusercontent.com/u/58836760?s=460&v=4
+---
+
+As library authors, we seek to provide the best possible developer experience for our users. Reducing time-to-error, narrowing down exactly what can be passed and accessed where, and providing intuitive APIs at every step all contribute to this. TypeScript has become a major tool in enabling us to ship such DX.
+
+However, as TypeScript usage becomes more widespread and libraries like tRPC and, more recently, [TanStack Router](https://tanstack.com/router/v1) are built entirely around it, we must understand the implications on type-checking and language server performance.
+
+## Where to start?
+
+First, consider if you need to optimize your types. App developers should not have to worry about types on that level, but library maintainers play a crucial role in ensuring their editing experience remains smooth.
+
+Regardless of whether your library is slow _now_, it's important to keep an eye on performance as your library changes. You must test your library's performance to the max **on each commit**.
+
+For tRPC, we ensure this by [generating](https://github.com/trpc/trpc/blob/9fc2d06a8924da73e10b9d4497f3a1f53de706ed/scripts/generateBigBoi.ts) and [testing](https://github.com/trpc/trpc/blob/9fc2d06a8924da73e10b9d4497f3a1f53de706ed/packages/tests/server/react/bigBoi.test.tsx) a router with 3,500 procedures and 1,000 routers. We test all aspects of the library: server, vanilla client, and the React client because they all have different code (type?) paths and have individually regressed in the past.
+
+All of these are indicators that you need to optimize your types:
+
+- Your library is slow to type-check (pretty straightforward)
+- Your library is slow to start up
+- Your library is slow to edit e.g. when you're typing in a file
+  - this is the most important one as it has the greatest effect on DX, you **never** want to make your users to wait for their editor to respond to a change
+
+## How to optimize?
+
+First, cut down as much as possible. Do you need the best possible TypeScript types everywhere? Is there anywhere you can cut down on the complexity of your types? Not every type needs to be perfect. The more complex your types are, the more work TypeScript has to do.
+
+example:
+
+do you really need this:
+
+```ts
+export const colorHexMap = {
+  red: '#ff0000',
+  green: '#00ff00',
+  blue: '#0000ff',
+  // ...1000 more colors
+} as const;
+
+const blueHex = colorHexMap.blue;
+```
+
+maybe this will work just as well:
+
+```ts
+export const colorHexMap: Record<string, string> = {
+  red: '#ff0000',
+  green: '#00ff00',
+  blue: '#0000ff',
+  // ...1000 more colors
+};
+
+const blueHex = colorHexMap.blue;
+```
+
+The second example is much simpler and faster to type-check.
+
+If you absolutely need your complex types, more work is required.
+
+### Tracing
```

Comment:
Maybe do this as a section that walk-through how ~~we~~you troubleshot this in tRPC

```suggestion
## A real-world example from tRPC

> PR in question: https://.....
```